File: frontend/src/types/track.ts
export type TrackStatus =
  | 'pending'
  | 'ok'
  | 'rejected'
  | 'duplicate'
  | 'needs_review'
  | 'error'
  | 'stale'

export type QualityTier = 'LOSSLESS' | 'HIGH' | 'MEDIUM' | 'LOW' | 'UNKNOWN'

export type TrackIssue =
  | 'missing_bpm'
  | 'missing_key'
  | 'missing_artist'
  | 'missing_title'
  | 'low_quality'
  | 'error'
  | 'needs_review'
  | 'weak_filename_parse'
  | 'suspicious_artist'
  | 'suspicious_title'

export type ParseConfidence = 'HIGH' | 'MEDIUM' | 'LOW' | 'UNKNOWN'

// Shape returned by GET /api/tracks (list view)
export interface TrackSummary {
  id:           number
  filepath:     string
  filename:     string
  artist:       string | null
  title:        string | null
  genre:        string | null
  bpm:          number | null
  key_camelot:  string | null
  key_musical:  string | null
  duration_sec: number | null
  bitrate_kbps: number | null
  status:       TrackStatus
  quality_tier: QualityTier | null
  parse_confidence: ParseConfidence | null
  issues:       TrackIssue[]
  recommended_action?: string | null
  recommended_route?: string | null
}

// Shape returned by GET /api/tracks/{id} (full detail)
export interface TrackDetail extends TrackSummary {
  filesize_bytes: number | null
  filesystem_path: string
  error_msg:      string | null
  processed_at:   string | null
  pipeline_ver:   string | null
  enrichment_queue_item?: Record<string, unknown> | null
}

export interface TrackStats {
  total:          number
  by_status:      Record<string, number>
  by_quality:     Record<string, number>
  missing_bpm:    number
  missing_key:    number
  missing_artist: number
  missing_title:  number
}

export interface TrackPage {
  items:  TrackSummary[]
  limit:  number
  offset: number
  total:  number
}

export interface TrackIssueCounts {
  missing_artist:       number
  missing_title:        number
  weak_filename_parse:  number
  suspicious_artist:    number
  suspicious_title:     number
}

export interface TrackListParams {
  path?:         string
  q?:            string
  search?:       string
  status?:       string
  artist?:       string
  genre?:        string
  key?:          string
  quality_tier?: string
  bpm_min?:      number
  bpm_max?:      number
  has_key?:      boolean
  issue?:        TrackIssue | string
  parse_confidence?: ParseConfidence | string
  sort?:         'artist' | 'title' | 'bpm' | 'processed_at' | 'filename'
  order?:        'asc' | 'desc'
  limit?:        number
  offset?:       number
}

export const ISSUE_LABELS: Record<TrackIssue, string> = {
  missing_bpm:    'No BPM',
  missing_key:    'No Key',
  missing_artist: 'No Artist',
  missing_title:  'No Title',
  low_quality:    'Low Quality',
  error:          'Error',
  needs_review:   'Needs Review',
  weak_filename_parse: 'Weak Filename Parse',
  suspicious_artist:   'Suspicious Artist',
  suspicious_title:    'Suspicious Title',
}

export const QUALITY_ORDER: QualityTier[] = ['LOSSLESS', 'HIGH', 'MEDIUM', 'LOW', 'UNKNOWN']

File: backend/app/schemas/track.py
"""
Pydantic schemas for the tracks API.

TrackSummary  — lightweight row shape used in list responses
TrackDetail   — full field set for single-track responses
TrackStats    — aggregate counts for the stats endpoint
TrackIssueItem — single item in the issues list response
"""
from __future__ import annotations

from typing import Dict, List, Optional

from pydantic import BaseModel

from ..models.track import Track


class TrackSummary(BaseModel):
    """Lightweight representation used in list/table responses."""

    id:           int
    filepath:     str
    filename:     str
    artist:       Optional[str] = None
    title:        Optional[str] = None
    genre:        Optional[str] = None
    bpm:          Optional[float] = None
    key_camelot:  Optional[str] = None
    key_musical:  Optional[str] = None
    duration_sec: Optional[float] = None
    bitrate_kbps: Optional[int] = None
    status:       str
    quality_tier: Optional[str] = None
    issues:       List[str] = []

    @classmethod
    def from_track(cls, t: Track) -> "TrackSummary":
        return cls(
            id=t.id,
            filepath=t.filepath,
            filename=t.filename,
            artist=t.artist,
            title=t.title,
            genre=t.genre,
            bpm=t.bpm,
            key_camelot=t.key_camelot,
            key_musical=t.key_musical,
            duration_sec=t.duration_sec,
            bitrate_kbps=t.bitrate_kbps,
            status=t.status,
            quality_tier=t.quality_tier,
            issues=t.issues,
        )


class TrackDetail(BaseModel):
    """Full field set returned for a single track."""

    id:             int
    filepath:       str
    filename:       str
    artist:         Optional[str] = None
    title:          Optional[str] = None
    genre:          Optional[str] = None
    bpm:            Optional[float] = None
    key_camelot:    Optional[str] = None
    key_musical:    Optional[str] = None
    duration_sec:   Optional[float] = None
    bitrate_kbps:   Optional[int] = None
    filesize_bytes: Optional[int] = None
    status:         str
    error_msg:      Optional[str] = None
    processed_at:   Optional[str] = None
    pipeline_ver:   Optional[str] = None
    quality_tier:   Optional[str] = None
    issues:         List[str] = []

    @classmethod
    def from_track(cls, t: Track) -> "TrackDetail":
        return cls(
            id=t.id,
            filepath=t.filepath,
            filename=t.filename,
            artist=t.artist,
            title=t.title,
            genre=t.genre,
            bpm=t.bpm,
            key_camelot=t.key_camelot,
            key_musical=t.key_musical,
            duration_sec=t.duration_sec,
            bitrate_kbps=t.bitrate_kbps,
            filesize_bytes=t.filesize_bytes,
            status=t.status,
            error_msg=t.error_msg,
            processed_at=t.processed_at,
            pipeline_ver=t.pipeline_ver,
            quality_tier=t.quality_tier,
            issues=t.issues,
        )


class TrackStats(BaseModel):
    """Aggregate counts returned by GET /api/tracks/stats."""

    total:          int
    by_status:      Dict[str, int]
    by_quality:     Dict[str, int]
    missing_bpm:    int
    missing_key:    int
    missing_artist: int
    missing_title:  int


class TrackIssueItem(BaseModel):
    """One entry in the issues list."""

    id:      int
    filepath: str
    filename: str
    artist:  Optional[str] = None
    title:   Optional[str] = None
    status:  str
    issues:  List[str]
